docs/design-doc.md

```diff
@@ -0,0 +1,296 @@
+# [Proposal] DWN Relay Service Design Doc  <!-- omit in toc -->
+
+# Contents <!-- omit in toc -->
+
+- [Abstract](#abstract)
+- [Handler Registration](#handler-registration)
+  - [Config file](#config-file)
+- [DWN Relay Message Handling](#dwn-relay-message-handling)
+  - [Example Message](#example-message)
+- [`DWN Relay -> Downstream Service` Request Payload](#dwn-relay---downstream-service-request-payload)
+  - [Example](#example)
+    - [DWeb Message](#dweb-message)
+    - [Downstream Request Body](#downstream-request-body)
+- [`Downstream Service -> DWN Relay` Response Payload](#downstream-service---dwn-relay-response-payload)
+  - [Example](#example-1)
+    - [Config](#config)
+    - [Response from Downstream Service](#response-from-downstream-service)
+- [`DWN Relay -> Client` Response Payload](#dwn-relay---client-response-payload)
+- [DID / Key management](#did--key-management)
+- [Additional Features](#additional-features)
+- [Configuration](#configuration)
+
+# Abstract
+The primary motivation behind this project is to streamline the ability for traditional back-end services that expose RESTful APIs to consume DWN messages (a.k.a DWeb Messages). The default behavior of the Relay Service is designed to prevent the leakage of DWN awareness or concepts into a downstream service.
+
+The Relay service runs as a standalone process that is separated by a network boundary. Inbound DWeb Messages are received by the relay service, integrity checked, and then relayed (aka proxied) to the downstream service.
+
+Integration between the relay service and the downstream service is achieved by registering (a.k.a configuring) a _Handler_. A _Handler_ is a mapping between a specific set of DWN message properties and an HTTP API endpoint exposed by the downstream service. e.g. "When a DWN message is received that matches the filter: `{ "method": "CollectionsWrite", "schema": "https://tbdex.io/schemas/Ask", "protocol": "tbdex" }` send it to `POST https://somehost.com/some-endpoint`". Handlers are configured via a config file that is loaded by the relay service on startup.
```

Comment:
Changing a handler would require service restart, is that a problem for steelthread?

We will probably add functionality to do it on the fly later

Reply:
dynamic config would be nice, but not necessary for a v1 IMO
whether through a config service, parameter store (like ssm), or plain old changing env vars that a process in the service can become aware of updates to